### PR TITLE
Removing 'http challenge' authentication flow and related authenticators

### LINKED ISSUE #20497: Move http-challenge authentication flow and the related authenticators to the extension
### Description

- Remove `http-challenge` authentication flow and related authenticators implementation and possibly move them to the extension https://github.com/keycloak-extensions/keycloak-openshift-ext
- We should make sure that migration works OK and there are not any errors when someone migrates from previous version (Possibly `http-challenge` flow should be removed during DB migration and also during JSON migration - maybe with usage of the `MigrateTo22_0_0` class)
- We should make sure that it's documented in the "Upgrading guide"
- We can try to make it tested inside "extension"

### Discussion

_No response_

### Motivation

Get rid of Openshift integration stuff, which is not actively tested in Keycloak

### Details

_No response_

### PR DESCRIPTION
**Pull request by mposolda**
closes #20497

Related PR to openshift extension: https://github.com/keycloak-extensions/keycloak-openshift-ext/pull/2

### CODE REVIEW

**Comment by mposolda** [comment]
@pedroigor Are you please able to review when you have a chance?
@hmlnarik Are you please able to review the parts related to migration (removal of `http challenge` flow during migration) or delegate to someone from the store team?

File: model/legacy-private/src/main/java/org/keycloak/migration/migrators/MigrateTo22_0_0.java
Comment (pedroigor) [inline model/legacy-private/src/main/java/org/keycloak/migration/migrators/MigrateTo22_0_0.java:59]: Doesn't it make more sense (and safer) to check if the authenticator is used within the flows rather than relying on the alias?

File: model/legacy-private/src/main/java/org/keycloak/migration/migrators/MigrateTo22_0_0.java
Comment (pedroigor) [inline model/legacy-private/src/main/java/org/keycloak/migration/migrators/MigrateTo22_0_0.java:65]: I'm wondering if we should also consider failing to start the server when we can't migrate the database automatically but it needs manual intervention.

I think this will make it very explicit that you can't upgrade until you change your realms manually.

For instance, in this case, we really don't want to have the `openshift-extension` mentioned and recommend deploying it. We know that extension is outdated and its current state is unknown.

**Review by pedroigor — CHANGES_REQUESTED** [review]

File: model/legacy-private/src/main/java/org/keycloak/migration/migrators/MigrateTo22_0_0.java
Comment (mposolda) [reply model/legacy-private/src/main/java/org/keycloak/migration/migrators/MigrateTo22_0_0.java:59]: The `Authentication Options` was subflow of the `Http Challenge` builtin flow. I cannot check for it's use with `KeycloakModelUtils.isFlowUsed` as it is just subflow, but I am checking that it's "builtin", which I supposed to be sufficient indication that it's the subflow of the `Http Challenge`.

I was considering another option - basically re-use same algorithm as used in `AuthenticationManagementResource.deleteFlow` to recursively delete the `Http Challenge` flow including it's subflow. Was about to move that stuff to `KEycloakModelUtils` to be able to use it from both services and migration code.

The small disadvantage of it are likely more DB queries as it needs to query executions etc, which looks like unnecessary DB queries during migration - especially in deployments with many realms. As we basically know that builtin flow `Http Challenge` has only 1 subflow called `Authentication Options`.

On the other hand, it's more proper approach to use `deleteFlow`, so I can change it if you prefer.

**Comment by mposolda** [comment]
@pedroigor Updated PR per your feedback. I've used separate commit for easier review (but I am planning to squash before merging).

File: model/legacy-private/src/main/java/org/keycloak/migration/migrators/MigrateTo22_0_0.java
Comment (pedroigor) [reply model/legacy-private/src/main/java/org/keycloak/migration/migrators/MigrateTo22_0_0.java]: TBH, I'm wondering if we should have migration at all but just include release notes.

We are adding code to something that is probably not used at all in the wild. And if so, it is supposed to break and force manual intervention based on the steps from the release notes.

Wdyt? Should be far simpler and less code to maintain and support.

**Review by pedroigor — DISMISSED** [review]
My comment about not doing migration is optional. LGTM if you want to ignore it.

File: model/legacy-private/src/main/java/org/keycloak/migration/migrators/MigrateTo22_0_0.java
Comment (mposolda) [reply model/legacy-private/src/main/java/org/keycloak/migration/migrators/MigrateTo22_0_0.java]: Well, I think that we definitely should have migration as otherwise people migrating from previous version will have `Http Challenge` flow in their realm, which would be just broken (as corresponding authenticators doesn't exists) and after click on this flow in the admin console, the error page would be shown. This looks like quite a buggy behaviour. The unused mess should be just removed IMO :-)

File: docs/documentation/upgrading/topics/keycloak/changes-22_0_0.adoc
Comment (andymunro) [inline docs/documentation/upgrading/topics/keycloak/changes-22_0_0.adoc:251]:
```suggestion
The built-in authentication flow `http challenge` was removed along with the authenticator implementations `no-cookie-redirect`, `basic-auth` and `basic-auth-otp`.
```

File: docs/documentation/upgrading/topics/keycloak/changes-22_0_0.adoc
Comment (andymunro) [inline docs/documentation/upgrading/topics/keycloak/changes-22_0_0.adoc:253]:
```suggestion
Authenticator implementations were moved to the Openshift extension described in the previous paragraph.
```

File: docs/documentation/upgrading/topics/keycloak/changes-22_0_0.adoc
Comment (andymunro) [inline docs/documentation/upgrading/topics/keycloak/changes-22_0_0.adoc:255]:
```suggestion
If you use the `http challenge` flow as a realm flow or as `First Broker Login` or `Post Broker Login` flow for any of your identity providers, the migration is not possible. Be sure to update
```

File: docs/documentation/upgrading/topics/keycloak/changes-22_0_0.adoc
Comment (andymunro) [inline docs/documentation/upgrading/topics/keycloak/changes-22_0_0.adoc:258]:
```suggestion
If you use the `http challenge` flow  as `Authentication Flow Binding Override` for any client, the migration would complete, but you could no longer log in to that client.
After the migration,  you would need to re-create the flow and update the configuration of your clients to use the new/differentJson flow.
```

**Review by andymunro — CHANGES_REQUESTED** [review]
@mposolda A few changes to suggest.

**Review by andymunro — CHANGES_REQUESTED** [review]
Looks good. A few changes to suggest, @mposolda.

**Review by hmlnarik — DISMISSED** [review]
The migration LGTM albeit it will be slow, but AFAICT, there is no easy way to improve that apart from no-downtime upgrade.

**Comment by mposolda** [comment]
@hmlnarik Yes, exactly. Thanks!

**Comment by mposolda** [comment]
@andymunro I've applied your suggestions to the documentation. Thanks!

**Pull request merged**
